### PR TITLE
Removed rescue block from `ActiveSupport::Cache::RedisCacheStore#handle_exception`

### PR DESCRIPTION
**Pull request by nstuart-at-salesforce**
...  to allow errors to be raised instead of treated as cache miss

### Summary

Issue #43082 was opened, as well as #42818. Connection errors, such as stale connections in the pool, can result in errors that are swallowed and treated as a cache miss. While there are parameters to handle this particular use case, `reconnect_attempts`, other cases were the cache-miss behavior happens can result in problematic behavior.

### CODE REVIEW

File: activesupport/lib/active_support/cache/redis_cache_store.rb [hunk 1]
```diff
@@ -140,11 +140,12 @@ def build_redis_client(url:, **redis_options)
       # Race condition TTL is not set by default. This can be used to avoid
       # "thundering herd" cache writes when hot cache entries are expired.
       # See <tt>ActiveSupport::Cache::Store#fetch</tt> for more.
-      def initialize(namespace: nil, compress: true, compress_threshold: 1.kilobyte, coder: default_coder, expires_in: nil, race_condition_ttl: nil, error_handler: DEFAULT_ERROR_HANDLER, **redis_options)
+      def initialize(namespace: nil, compress: true, compress_threshold: 1.kilobyte, coder: default_coder, expires_in: nil, race_condition_ttl: nil, error_handler: DEFAULT_ERROR_HANDLER, raise_errors: false, **redis_options)
```
Comment (nstuart-at-salesforce) [inline activesupport/lib/active_support/cache/redis_cache_store.rb:143]: `raise_errors` seems generic, but I didn't want to get overly verbose either. Open to suggestions

File: activesupport/lib/active_support/cache/redis_cache_store.rb [hunk 2]
```diff
@@ -460,7 +461,12 @@ def failsafe(method, returning: nil)
           yield
         rescue ::Redis::BaseError => e
           handle_exception exception: e, method: method, returning: returning
-          returning
+
+          if @raise_errors
```
Comment (nstuart-at-salesforce) [inline activesupport/lib/active_support/cache/redis_cache_store.rb:465]: Still call `handle_exception` to allow for global handlers as well as local ones(rescue).

File: activesupport/test/cache/behaviors/failure_raising_behavior.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+module FailureRaisingBehavior
```
Comment (nstuart-at-salesforce) [inline activesupport/test/cache/behaviors/failure_raising_behavior.rb:3]: Same as `FailureSuccessBehavior`, except it doesn't assert values

File: activesupport/test/cache/behaviors/failure_safety_behavior.rb [hunk 4]
```diff
@@ -9,7 +9,17 @@ def test_fetch_read_failure_returns_nil
     end
   end
 
-  def test_fetch_read_failure_does_not_attempt_to_write
```
Comment (nstuart-at-salesforce) [inline activesupport/test/cache/behaviors/failure_safety_behavior.rb:12]: I saw this was empty, so I wrote a test in the spirit of the name.

**Comment by nstuart-at-salesforce** [comment]
> I think all this would be way simpler if instead of adding an extra `raise_errors` option, you'd sligthly modify the existing `error_handler`, so that it have the option to raise. e.g.:
> 
> ```ruby
>         def handle_exception(exception:, method:, returning:)
>           if @error_handler
>             @error_handler.(method: method, exception: exception, returning: returning)
>           end
>         rescue => failsafe
>           warn "RedisCacheStore ignored exception in handle_exception: #{failsafe.class}: #{failsafe.message}\n  #{failsafe.backtrace.join("\n  ")}"
>         end
> ```
> 
> would become:
> 
> ```ruby
>         def handle_exception(exception:, method:, returning:)
>           if @error_handler
>             @error_handler.(method: method, exception: exception, returning: returning)
>           end
>         end
> ```
> 
> It's much more flexible, and doesn't introduce a new option.

@byroot 
Agreed it does not introduce a new option, but it does change the behavior of the `error_handler`, which now can throw an error that was previously expected to be rescued. I like it, I was worried about backward compatibility breaking with new logic such as that.

**Comment by byroot** [comment]
> but it does change the behavior of the `error_handler`, which now can throw an error that was previously expected to be rescued.

Yeah I know. But I think it's fine. It's a very fringe option, and it's easy for the rare people relying on this to just put the `rescue` inside their handler.

File: guides/source/caching_with_rails.md [hunk 5]
```diff
@@ -528,7 +528,6 @@ config.cache_store = :redis_cache_store, { url: cache_servers,
   read_timeout:       0.2, # Defaults to 1 second
   write_timeout:      0.2, # Defaults to 1 second
   reconnect_attempts: 1,   # Defaults to 0
-
```
Comment (ghiculescu) [inline guides/source/caching_with_rails.md:531]: is this change needed?

File: activesupport/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,9 @@
+*   Removed rescue block from `ActiveSupport::Cache::RedisCacheStore#handle_exception`
+
+    Previously, if you provided a `error_handler` to `redis_cache_store`, any errors thrown by
+    the error handler would be rescued and logged only. Removed the `rescue` clause from `handle_exception`
+    to allow these to be thrown.
```
Comment (ghiculescu) [inline activesupport/CHANGELOG.md:5]: add your name here

File: activesupport/test/cache/stores/redis_cache_store_test.rb [hunk 7]
```diff
@@ -283,6 +283,40 @@ def ensure_connected
...
 
+  class FailureRaisingFromUnavailableClientTest < StoreTest
+    include FailureRaisingBehavior
+
+    private
+      def emulating_unavailability
+        old_client = Redis.send(:remove_const, :Client)
+        Redis.const_set(:Client, UnavailableRedisClient)
+
+        assert_raise Redis::BaseConnectionError do
```
Comment (byroot) [inline activesupport/test/cache/stores/redis_cache_store_test.rb:294]: I don't think `assert_raise` should be part of `emulating_unavailability`, because it makes it very unclear in the calling test.

When I was reading them I was wondering why no error was raised even thought it was the goal of your PR.

Comment (ghiculescu) [reply activesupport/test/cache/stores/redis_cache_store_test.rb:294]: Haha I just wrote the exact same comment ^

**Comment by rafaelfranca** [comment]
Remember to check the commit history. There is a commit in this PR with the title "init commit" that doesn't match the change.

**Comment by byroot** [comment]
Indeed, sorry about that, I totally forgot in this case. I should have asked to squash it all.
